### PR TITLE
Support for Yarn Berry

### PR DESCRIPTION
**Pull request by thitch97**
[Readable](https://github.com/paketo-buildpacks/rfcs/blob/yarn2/text/nodejs/0014-yarn-berry.md)

## Checklist

* [ ] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [ ] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: text/nodejs/0014-yarn-berry.md
Comment (robdimsdale) [inline text/nodejs/0014-yarn-berry.md:178]: Is it possible to have a `.pnp.cjs` file but no `.yarn/cache`? What about vice-versa - `.yarn/cache` but no `.pnp.cjs`? What should the buildpack do in those cases?

File: text/nodejs/0014-yarn-berry.md
Comment (robdimsdale) [inline text/nodejs/0014-yarn-berry.md:159]: Will `Build` create a filesystem-based (i.e. syft-based) SBOM like `yarn-install` does today?

File: text/nodejs/0014-yarn-berry.md
Comment (robdimsdale) [inline text/nodejs/0014-yarn-berry.md:45]: Will we have to validate that `yarn` is vendored under `.yarn/releases` - either during `Detect` or during `Build`? Some of the workflows below invoke `yarn` so if users do not provide it under `.yarn/releases` and this buildpack does not require the `yarn` buildpack then those workflows will fail.

Perhaps more concretely, is it possible to pass `Detect` (by validating that `.yarnrc.yml` exists) and then fail at run time because `.yarn/releases` does not exist?

File: text/nodejs/0014-yarn-berry.md
Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md:178]: In both cases, `yarn install`  would need to be run. The first case is implied above but could admittedly be made more explicit. I’ll update to include the second case as well.

File: text/nodejs/0014-yarn-berry.md
Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md:159]: Good question. Yes, syft [supports yarn berry](https://github.com/anchore/syft/pull/868). That should be called out here explicitly.

File: text/nodejs/0014-yarn-berry.md
Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md:45]: The build would fail since `corepack` (which manages per-project installations of yarn & comes bundled with Node) won’t be able to locate the release as specified by `.yarnrc.yml`.  This raises additional questions and there will be challenges supporting offline builds with this approach. I’ll outline those issues in `Unresolved Questions`.

File: text/nodejs/0014-yarn-berry.md
Comment (ryanmoran) [inline text/nodejs/0014-yarn-berry.md:42]: Why does this require anything at launch? At launch, this buildpack isn't really doing anything other than making sure that there are some JavaScript files available on the filesystem. I'd expect that if someone wanted to run these files, they'd need to require `node` some other way, no?

File: text/nodejs/0014-yarn-berry.md
Comment (ryanmoran) [reply text/nodejs/0014-yarn-berry.md:177]: Can you clarify what happens if there is a `.pnp.cjs` file and a local cache and the `yarn.lock` file has changed?

**Review by ryanmoran — CHANGES_REQUESTED** [review]

File: text/nodejs/0014-yarn-berry.md
Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md]: You are correct, I've updated accordingly.

File: text/nodejs/0014-yarn-berry.md
Comment (aurelienbottazini) [inline text/nodejs/0014-yarn-berry.md:209]: I am bit concerned by this.
I agree `--immutable` is a good idea in theory.
But I had strange cases where for an identical node and yarn version, `yarn install --immutable` would fail on some env and succeed on others. Very frustrating and hard to debug.

Should we instead not specify the option by default and let users either
- add `enableImmutableInstalls: false` to `.yarnrc.yml`
- or set with yarn env vars mechanism: `YARN_ENABLE_IMMUTABLE_INSTALLS`

Comment (aurelienbottazini) [inline text/nodejs/0014-yarn-berry.md:209]: I am bit concerned by this.
I agree `--immutable` is a good idea in theory.
But I had strange cases where for an identical node and yarn version, `yarn install --immutable` would fail on some env and succeed on others. Very frustrating and hard to debug.

Should we instead not specify the option by default and let users either
- add `enableImmutableInstalls: false` to `.yarnrc.yml`
- or set with yarn env vars mechanism: `YARN_ENABLE_IMMUTABLE_INSTALLS`

Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md:209]: This is helpful feedback, thank you. I'm okay with making this configurable.

File: text/nodejs/0014-yarn-berry.md
Comment (ryanmoran) [inline text/nodejs/0014-yarn-berry.md:246]: I think it might be reasonable to still get Berry support in the existing buildpack if we just take some effort to keep the "Classic" and Berry parts separate. If we did something like move all the existing code in the buildpack into a `classic` package and then created a new `berry` package we can keep them logically separated. We could then just have a small amount of code at the top-level that decides with codepath to invoke.

Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md:246]: I hadn't considered separate packages, that's an interesting suggestion. It would remove a lot of overhead in the event that we drop support for Classic altogether.

File: text/nodejs/0014-yarn-berry.md
Comment (ekcasey) [inline text/nodejs/0014-yarn-berry.md:230]: Can we elaborate on any changes needed to the yarn-start or node-start (included in the berry group above) buildpacks? Seems like there could be some gotchas here with detection criteria and cache vs modules configurations.

File: text/nodejs/0014-yarn-berry.md
Comment (ekcasey) [inline text/nodejs/0014-yarn-berry.md:71]: Should this be optional and opt out if users are going with the [zero-installs](https://yarnpkg.com/features/zero-installs) method? This might mess up detection for the group without some changes in (yarn start), but it could be confusing to see yarn install opt-in when you are explicitly striving for zero-installs?

File: text/nodejs/0014-yarn-berry.md
Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md:74]: Good question. I think the inclusion of `node-module-bom` here is a mistake on my part. The buildpack shouldn't need to leverage that buildpack at all for generating the SBoM and would instead include logic for SBoM generation akin to what [exists](https://github.com/paketo-buildpacks/yarn-install/blob/b28f4b300c67bf5bb3c7a55e06cc782f24ed0dee/build.go#L240) in `yarn-install` currently.

File: text/nodejs/0014-yarn-berry.md
Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md:71]: This order group may now be obsolete as I was able to prove out a sane way to keep the build logic for both Yarn classic and Berry in a single buildpack (per @ryanmoran's suggestion). As for the UX, I figured we could indicate through the logs whenever the buildpack has chosen to forgo running `yarn install` based on adherence to the zero-installs setup.

**Comment by divoxx** [comment]
Is this RFC still actively being worked on? We have interest in helping push this forward as well, if help is needed.

**Comment by ekcasey** [comment]
@thitch97 IIRC you mentioned during WG you were making some largish changes to the RFC. Do you mind posting a comment describing where it is at and what your intentions are? If it needs more work before it is ready for further discussion/reviews maybe we should put it back into draft temporarily?

**Comment by thitch97** [comment]
@divoxx @ekcasey My intention at the moment is to shift to the model initially enumerated in the "Rationale and Alternatives" section. It turns out that hosting both Classic and Berry codepaths in one buildpack isn't as heavy an engineering/maintenance burden as it seemed. As such, many of the technical decisions here, especially the order groupings, are now obsolete.

**Comment by fg-j** [comment]
From WG 7/19/22:

- In latest update of this RFC, the Build Plan dependency `node_modules` has been renamed `node_packages`. This simplifies plan resolution.
- In latest update, Yarn Berry builds still require `yarn` at build, instead of using `corepack`. This presents some issues that are hashed out in the Alternatives and Unresolved Questions sections.
- Concern: Yarn Berry doesn't seem to play well with offline builds. 
  - Comment: "offline" builds are rare; builds that only have access to a private network are more common
  - Comment: there still is a fully offline workflow for yarn 1
  - Q: Could we enable users to volume-mount a package cache for offline builds? The maven buildpack currently supports this?
    -  This sounds tricky but possible. In yarn context, your global cache on your workstation might contain dependencies that don't have anything to do with the app in question. Also, the location of the package cache isn't configurable for yarn.

- **Requested change:** Add a list of buildpacks that provide or require the `node_modules` plan entry so it's easy to tell which buildpacks would need to change 

Seeking @paketo-buildpacks/nodejs-maintainers  feedback on these changes.

File: text/nodejs/0014-yarn-berry.md
Comment (ryanmoran) [inline text/nodejs/0014-yarn-berry.md:62]: Can you provide some more of the rational behind this choice? I'm not seeing how we don't end up in a state where its either confusing to Berry users or confusing to non-Berry users about what is being provided here.

Comment (fg-j) [reply text/nodejs/0014-yarn-berry.md:62]: To me, it'd be more intuitive if the yarn classic and yarn berry paths resulted in two slightly different build plans, since they do involve dealing with different sets of artifacts. 

I'd be in favor of the yarn-install buildpack providing `node_modules` if it plans on running `yarn install` and `yarn_cache` if it doesn't. Then the start command buildpacks could require one `Or` the other.

Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md:62]: @ryanmoran does @fg-j's comment more closely align with what you'd like to see? I'm fine with additional logic for the sake of clarity.

**Comment by joshuatcasey** [comment]
2022-07-26 Working Group notes:

File: text/nodejs/0014-yarn-berry.md
Comment (ryanmoran) [inline text/nodejs/0014-yarn-berry.md:62]: I don't think this matches the described API below anymore.

Comment (thitch97) [reply text/nodejs/0014-yarn-berry.md]: @ryanmoran good catch. I've updated that now.

**Review by ryanmoran — APPROVED** [review]

**Comment by dmikusa-pivotal** [comment]
2022-08-02 WG Notes:

Approved by Node maintainers. Final call for any comments. Scheduled to be merged soon.

**Pull request merged**
